### ISSUE TITLE
weird behavior in distributed training using partitioner

### ISSUE THREAD

**Issue by Zhiqiang-Ye** [tensorflow/tensorflow#24612]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux CentOS 7.3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): unknown, probably binary
- TensorFlow version (use command below): 1.10.0
- Python version: 2.7.5
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: NA
- GPU model and memory: NA

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**

I wrote a custom estimator to implement FM algorithm and used train_and_evaluate to perform distributed training (4 ps and 100 workers). When I set partitioner to None, the training process quickly converged to an expected test AUC 0.87. However, when I either used fixed_size_partitioner or min_max_variable_partitioner with modified min_slice_size, the test AUC stayed below 0.64.  Different learning rates and batch sizes didn't help. Am I missing anything in distributed TF?

**Describe the expected behavior**

With partitioner dividing embedding matrix to each PS, the training process should also converge quickly.

**Code to reproduce the issue**
```
class FMEstimator:
	def __init__(self, model_dir, config=None, params=None, optimizer=None, partitioner=None):
		self.model_dir = model_dir
		self.config = config
		self.params = params
		self.optimizer = optimizer
		self.partitioner = partitioner

	def __embedding_lookup_square_sparse(self, params, sp_ids, partition_strategy="mod", name=None, max_norm=None):
		if isinstance(params, variables.PartitionedVariable):
			params = list(params)  # Iterate to get the underlying Variables.
		if not isinstance(params, list):
			params = [params]

		with ops.name_scope(name, "embedding_lookup_square_sparse",
		                    params + [sp_ids]) as name:
			segment_ids = sp_ids.indices[:, 0]
			if segment_ids.dtype != dtypes.int32:
				segment_ids = math_ops.cast(segment_ids, dtypes.int32)

			ids = sp_ids.values
			ids, idx = array_ops.unique(ids)

			embeddings = tf.nn.embedding_lookup(params, ids, partition_strategy=partition_strategy, max_norm=max_norm)

			embeddings = tf.square(embeddings)

			assert idx is not None

			embeddings = math_ops.sparse_segment_sum(embeddings, idx, segment_ids, name=name)

			return embeddings

	def get_model_fn(self):
		def custom_model_fn(features, labels, mode, params):
			linear_bias = tf.get_variable(name='linear_bias',
			                              shape=[1],
			                              dtype=tf.float32,
			                              initializer=tf.random_normal_initializer(stddev=0.0001))

			linear_w = tf.get_variable(name='linear_w',
			                           shape=[params['feature_size'], 1],
			                           dtype=tf.float32,
			                           initializer=tf.random_normal_initializer(stddev=0.0001),
			                           partitioner=self.partitioner)

			# wx
			# size: [batch_size, 1]
			logits_wide = tf.nn.embedding_lookup_sparse(params=linear_w,
			                                            sp_ids=features['featureID'],
			                                            sp_weights=None,
			                                            combiner='sum')
			# wx + b
			logits_linear = linear_bias + logits_wide

			cross_emb_w = tf.get_variable(name='cross_emb_w',
			                              shape=[params['feature_size'], params['cross_emb_size']],
			                              dtype=tf.float32,
                                                      partitioner=self.partitioner,
			                              initializer=tf.random_normal_initializer(stddev=0.0001))

			# (a1,b1,c1) (a2,b2,c2) -> (a1+a2, b1+b2, c1+c2)
			# size: [batch_size, cross_emb_size]
			summed_cross_emb = tf.nn.embedding_lookup_sparse(params=cross_emb_w,
			                                                 sp_ids=features['featureID'],
			                                                 sp_weights=None,
			                                                 combiner='sum')
			# ((a1+a2)^2, (b1+b2)^2, (c1+c2)^2)
			# size: [batch_size, cross_emb_size]
			squared_summed_cross_emb = tf.square(summed_cross_emb)

			# (a1^2, b1^2, c1^2) (a2^2, b2^2, c2^2) -> (a1^2+a2^2, b1^2+b2^2, c1^2+c2^2)
			# size: [batch_size, cross_emb_size]
			summed_squared_cross_emb = self.__embedding_lookup_square_sparse(params=cross_emb_w,
			                                                                 sp_ids=features['featureID'])

			# a1a2+b1b2+c1c2
			# size: [batch_size, 1]
			logits_cross = tf.reduce_sum(0.5 * tf.subtract(squared_summed_cross_emb, summed_squared_cross_emb), 1, keepdims=True)

			# logits = tf.add(logits_linear, logits_cross)
			logits = logits_linear
			logits_adjusted = logits + tf.math.log(params['negative_sampling_rate'])

			if mode == tf.estimator.ModeKeys.PREDICT:
				predictions = {
					'probabilities': tf.nn.sigmoid(logits_adjusted),
					'logits': logits,
					'logits_adjusted': logits_adjusted
				}

				return tf.estimator.EstimatorSpec(mode, predictions=predictions)

			else:
				loss = tf.reduce_mean(
					tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.cast(labels, dtype=tf.float32),
					                                        logits=logits))

				if mode == tf.estimator.ModeKeys.EVAL:
					auc = tf.metrics.auc(
						labels=labels,
						predictions=1 / (1 + tf.math.exp(-logits_adjusted)),
						num_thresholds=400,
						curve='ROC',
						summation_method='careful_interpolation')
					logloss = tf.metrics.mean(tf.nn.sigmoid_cross_entropy_with_logits(
						labels=tf.cast(labels, dtype=tf.float32),
						logits=logits_adjusted))
					tf.summary.scalar('True_AUC', auc)
					tf.summary.scalar('True_Logloss', logloss)
					metrics = {
						'True_AUC': auc,
						'True_Logloss': logloss
					}

					predictions = {
						'probabilities': tf.nn.sigmoid(logits_adjusted),
						'logits': logits,
						'logits_adjusted': logits_adjusted
					}

					return tf.estimator.EstimatorSpec(mode, loss=loss, predictions=predictions,
					                                  eval_metric_ops=metrics)

				elif mode == tf.estimator.ModeKeys.TRAIN:
					train_op = self.optimizer.minimize(loss, global_step=tf.train.get_global_step())

					return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)

		return custom_model_fn

	def get_estimator(self):
		return tf.estimator.Estimator(model_fn=self.get_model_fn(),
		                              model_dir=self.model_dir,
		                              config=self.config,
		                              params=self.params)


```
**Other info / logs**
The input data is in sparse tensor of tf.int64. Feature size is around 480,000. FM cross_emb_size is 4.
I have checked the graph from tensorboard and timeline from hooks. Both of them seemed correct. The embedding matrix was divided and placed on each ps with corresponding update ops. I used both the estimator.evaluate function and hand-made prediction calculation with exported embedding matrix, and got the same low AUC value. I also ask a question in stackoverflow ([stackoverflow_link](https://stackoverflow.com/questions/53870450/distributed-tf-multi-ps-version-lr-doesnt-converge)), but go no response by now.

**Comment by msymp** [comment]
Hi @karmel , this user is using the tf.estimator module to write a custom estimator and trying to set the Partitioner in the model. But it does not converge. Can you please see why this happens? Thanks.

**Comment by karmel** [comment]
@xiejw , @alextp for the behavior of the current partitioners, and some insight as to the expected APIs in 2.0.

**Comment by alextp** [comment]
The partitioner shouldn't affect convergence, other than it can change the relative speed of your workers and hence affect the level of parallelism requiring a retuning of the learning rate.

**Comment by Zhiqiang-Ye** [comment]
> 
> 
> The partitioner shouldn't affect convergence, other than it can change the relative speed of your workers and hence affect the level of parallelism requiring a retuning of the learning rate.

I had a similar guess about the parallelism and did some experiments under less scale setting using 2 ps and 5 workers on my local server. I also observed low AUC with fixed_size_partitioner. After reading your reply, I did another set of experiments with only 1 worker providing updating gradients, and still low AUC.

I also tried to find some other clues from TensorBoard and noticed:
1. the training average losses actually were similar whether using partitioner or not. The numbers fluctuated in the same region. That's why I tried several ways (estimator.evaluate, NewCheckpointReader and saver.restore) to do the evaluation cause I thought the training process with partitioner might acutally have converged and there might be problem in my evaluation process. But the results were the same of low AUC.

2. Since the training process with partitioner gave acceptable average loss, I used the exact training data to the evaluation and expected similar loss. However, large loss number came with low AUC. 

Currently, I decide to set evaluator in the cluster_spec to get runtime auc and loss of both training data and test data to see what will happen. If the runtime evalutor could give good AUC, I may have problem exporting weights, otherwise, I don't know what I can do...

**Comment by msymp** [comment]
Hello @Zhiqiang-Ye , we hope @alextp was able to address your initial query. We shall close this issue now. Thanks.

**Comment by Zhiqiang-Ye** [comment]
@alextp @msymp 
Sorry for my late reply. I just got time to do more tests and finally solve this issue.

On one hand, the training process went well with partitioner, and it converged quickly;
On the other hand, the discrepancy between evaluation AUC and training AUC resulted from the initialization process of variables in the graph. To get the expected result, I had to provide a partitioner with **exactly the same** partitioner args when using tf.get_variable to make these variables as partitionedVariable, even though the evaluation process can only run in the non-distributed mode. Simple tensor or partitionedVariable with wrong partitioner info would gave wrong result.

Do I miss any points in the doc about this initialization requirements, or is there any problem with my code about using partitioner? I didn't expect issues in this part, since there was no error when using saver.restore to initialize simple tensor variable and the tensor values were the same as their counterparts extracted from checkpoint.
